cpp/ql/src/Likely Bugs/OO/UnsafeUseOfThis.ql

```diff
@@ -0,0 +1,204 @@
+    // Weed out implicit calls to destructors of a base class
+    not func instanceof Destructor
+  )
+}
+
+/** Holds if `init` initializes the `this` pointer in class `c`. */
+predicate isSource(InitializeParameterInstruction init, string msg, Class c) {
+  (
+    exists(Constructor func |
+      not func instanceof CopyConstructor and
+      not func instanceof MoveConstructor and
+      func = init.getEnclosingFunction() and
+      msg = "construction"
+    )
+    or
+    init.getEnclosingFunction() instanceof Destructor and msg = "destruction"
+  ) and
+  init.getIRVariable() instanceof IRThisVariable and
+  init.getEnclosingFunction().getDeclaringType() = c
+}
+
+/** Holds if `instr` flows to a sink. */
+predicate flowsToSink(Instruction instr) {
+  isSink(instr, _, _)
+  or
+  exists(Instruction mid |
+    successor(instr, mid) and
+    flowsToSink(mid)
+  )
+}
+
+/** Holds if `instr` flows from a source. */
+predicate flowsFromSource(Instruction instr) {
+  isSource(instr, _, _)
+  or
+  exists(Instruction mid |
+    successor(mid, instr) and
+    flowsFromSource(mid)
+  )
+}
+
+/**
+ * Holds if `instr` is an argument (or argument indirection) to a call, and
+ * `succ` is the corresponding initialization instruction in the call target.
+ */
+predicate flowThroughCallable(Instruction instr, Instruction succ) {
+  flowsToSink(succ) and
+  (
+    // Flow from an argument to a parameter
+    exists(CallInstruction call, InitializeParameterInstruction init | init = succ |
+      getPositionalArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+      or
+      getThisArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+    )
+    or
+    // Flow from argument indirection to parameter indirection
+    exists(
+      CallInstruction call, ReadSideEffectInstruction read, InitializeIndirectionInstruction init
+    |
+      init = succ and
+      read.getPrimaryInstruction() = call and
+      init.getEnclosingFunction() = call.getStaticCallTarget()
+    |
+      exists(int n |
+        read.getSideEffectOperand().getAnyDef() = instr and
+        read.getIndex() = n and
+        init.getParameter().getIndex() = unbind(n)
+      )
+      or
+      call.getThisArgument() = instr and
+      init.getIRVariable() instanceof IRThisVariable
+    )
+  )
+}
+
+/** Holds if `instr` flows to `succ`. */
+predicate successor(Instruction instr, Instruction succ) {
+  flowsToSink(succ) and
+  (
+    irBbPostDominates(succ.getBlock(), instr.getBlock()) and
+    (
+      succ.(CopyInstruction).getSourceValue() = instr or
+      succ.(CheckedConvertOrNullInstruction).getUnary() = instr or
+      succ.(ChiInstruction).getTotal() = instr or
+      succ.(ConvertInstruction).getUnary() = instr or
+      succ.(InheritanceConversionInstruction).getUnary() = instr
+    )
+    or
+    flowThroughCallable(instr, succ)
+  )
+}
+
+predicate successorTC(Instruction i1, Instruction i2) = fastTC(successor/2)(i1, i2)
```

Comment:
```suggestion
predicate successorTC(Instruction i1, Instruction i2) { successor+(i1, i2) }
```
I hope this gives the same result as using `fastTC` explicitly. If not, I think the core team would like to know why.

Reply:
896dee5fb79c528400f082144d67bb76de27f273 replaces the `fastTC` with manual recursion.